File: factory/factory.js
class Point
{
  constructor(x, y)
  {
    this.x = x;
    this.y = y;
  }

  static get factory()
  {
    return new PointFactory();
  }
}

// Factory
class PointFactory
{
  // Factory Method
  static newCartesianPoint(x, y)
  {
    return new Point(x, y);
  }

  static newPolarPoint(rho, theta)
  {
    return new Point(
      rho * Math.cos(theta),
      rho * Math.sin(theta)
    )
  }
}

// Usage
let p1 = PointFactory.newCartesianPoint(4, 5);
console.log(p1);

let p2 = PointFactory.newPolarPoint(4, 5);
console.log(p2);

File: factory/factory.py
from enum import Enum
from math import *


class CoordinateSystem(Enum):
    CARTESIAN = 1
    POLAR = 2


class Point:
    def __init__(self, x=0, y=0):
        self.x = x
        self.y = y

    def __str__(self):
        return f'x: {self.x}, y: {self.y}'

    class PointFactory:
        @staticmethod
        def new_cartesian_point(x, y):
            return Point(x, y)

        @staticmethod
        def new_polar_point(rho, theta):
            return Point(rho * cos(theta), rho * sin(theta))

    factory = PointFactory()


if __name__ == '__main__':
    p = Point.factory.new_cartesian_point(2, 3)
    p2 = Point.factory.new_polar_point(1, 2)
    print(p, p2)
